app/models/search_suggestion.rb

```diff
@@ -38,15 +38,22 @@ def fuzzy_search(query)
           .limit(10)
           .all
       else
-        suggestions = where(id: distinct_values(query).from_self.select(:id))
-          .with_query(query)
-          .with_score(query)
-          .order(
-            Sequel.asc(:priority),
-            Sequel.desc(:score),
+        suggestions = with_sql("
+          WITH filtered AS (
+            SELECT DISTINCT ON (value) *
+            FROM search_suggestions
+            WHERE
+              type IN (?, ?) AND
+              value ILIKE ?
+            ORDER BY value, priority
           )
-          .limit(10)
-          .all
+          SELECT *,
+                ? AS query,
+                similarity(value, ?) AS score
+          FROM filtered
+          ORDER BY priority ASC, score DESC
+          LIMIT ?;
+        ", 'search_reference', 'full_chemical_name', "%#{query}%", query, query, 10).all
```

Comment:
I think there's a mismatch between the number of passed parameters and we're also bypassing the builtin sanitisation by going direct for sql :thinking:

Reply:
```ruby
        filtered_cte = select_all
          .distinct(:value)
          .where(type: ['search_reference', 'full_chemical_name'])
          .where(Sequel.ilike(:value, "%#{query}%"))
          .order(:value, :priority)

        suggestions = with(:filtered, filtered_cte)
          .from(:filtered)
          .select_all(:filtered)
          .with_query(query)
          .with_score(query)
          .order(:priority, Sequel.desc(:score))
          .limit(10)
          .all

        suggestions.select do |suggestion|
          # ilike filters can return suggestions with a score of 0
          suggestion[:score] > ILIKE_SIMILARITY_THRESHOLD
        end
```